components/brave_ads/content/browser/search_result_ad/search_result_ad_handler.cc

```diff
@@ -0,0 +1,112 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_ads/content/browser/search_result_ad/search_result_ad_handler.h"
+
+#include <utility>
+
+#include "base/memory/ptr_util.h"
+#include "brave/components/brave_ads/browser/ads_service.h"
+#include "brave/components/brave_ads/common/features.h"
+#include "brave/components/brave_ads/core/browser/search_result_ad/search_result_ad_converting_util.h"
+#include "brave/components/brave_search/common/brave_search_utils.h"
+#include "content/public/browser/render_frame_host.h"
+#include "services/service_manager/public/cpp/interface_provider.h"
+
+namespace brave_ads {
+
+SearchResultAdHandler::SearchResultAdHandler(AdsService* ads_service,
+                                             bool should_trigger_viewed_event)
+    : ads_service_(ads_service),
+      should_trigger_viewed_event_(should_trigger_viewed_event) {
+  DCHECK(ads_service_);
+}
+
+SearchResultAdHandler::~SearchResultAdHandler() = default;
+
+// static
+std::unique_ptr<SearchResultAdHandler>
+SearchResultAdHandler::MaybeCreateSearchResultAdHandler(
+    AdsService* ads_service,
+    const GURL& url,
+    bool should_trigger_viewed_event) {
+  if (!ads_service || !ads_service->IsEnabled() ||
+      !base::FeatureList::IsEnabled(
+          features::kSupportBraveSearchResultAdConfirmationEvents) ||
+      !brave_search::IsAllowedHost(url)) {
+    return {};
+  }
+
+  return base::WrapUnique(
+      new SearchResultAdHandler(ads_service, should_trigger_viewed_event));
+}
+
+void SearchResultAdHandler::MaybeRetrieveSearchResultAd(
+    content::RenderFrameHost* render_frame_host) {
+  DCHECK(render_frame_host);
+  DCHECK(ads_service_);
+
+  if (!ads_service_->IsEnabled()) {
```

Comment:
Is it possible that ads_service could be null, and we should fail gracefully?

Reply:
No, `ads_service_` cannot be null for `SearchResultAdHandler`. We have a `DCHECK(ads_service_)` in `SearchResultAdHandler` constructor.